Kernel/Devices/Generic/PCSpeakerDevice.cpp

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2023, Liav A. <liavalb@hotmail.co.il>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/API/ToneGenerationPacket.h>
+#if ARCH(X86_64)
+#    include <Kernel/Arch/x86_64/PCSpeaker.h>
+#endif
+#include <Kernel/Boot/CommandLine.h>
+#include <Kernel/Devices/DeviceManagement.h>
+#include <Kernel/Devices/Generic/PCSpeakerDevice.h>
+#include <Kernel/Sections.h>
+
+namespace Kernel {
+
+UNMAP_AFTER_INIT NonnullRefPtr<PCSpeakerDevice> PCSpeakerDevice::must_create()
+{
+    auto device = MUST(DeviceManagement::try_create_device<PCSpeakerDevice>());
+    return *device;
+}
+
+UNMAP_AFTER_INIT PCSpeakerDevice::PCSpeakerDevice()
+    : CharacterDevice(1, 10)
+{
+}
+
+UNMAP_AFTER_INIT PCSpeakerDevice::~PCSpeakerDevice() = default;
+
+bool PCSpeakerDevice::can_read(OpenFileDescription const&, u64) const
+{
+    return true;
+}
+
+ErrorOr<size_t> PCSpeakerDevice::read(OpenFileDescription&, u64, UserOrKernelBuffer&, size_t)
+{
+    return Error::from_errno(ENOTIMPL);
+}
+
+ErrorOr<size_t> PCSpeakerDevice::write(OpenFileDescription&, u64, UserOrKernelBuffer const& buffer, size_t buffer_size)
+{
+    ToneGenerationPacket packet {};
+    TRY(buffer.read(&packet, sizeof(ToneGenerationPacket)));
```

Comment:
We should reject the request before this in case `buffer_size < sizeof(ToneGenerationPacket)`, maybe even if `buffer_size % sizeof(ToneGenerationPacket) != 0`. Otherwise, we end up doing an out-of-bounds read.

The only case where we would notice an out-of-bounds read right now is if it was at the end of the allowed userspace memory, and then it would just return `EFAULT`, which is likely not what we want.

Reply:
In fact, our current non-use of `buffer_size` makes the build fail on non-x86-64 platforms.